R/functions.R

```diff
@@ -368,6 +398,115 @@ out_clique_decomposition = function(NEgraph,Zobs_id,minr,minc,stop_at_Zobs=TRUE)
+#' @export
+greedy_decom = function(ne, num_ass){
+  focal_units = c() # set of focal units
+  focal_ass = c() # set of focal assignments
+
+  CONT = TRUE
+  ne_original = ne
+  get_clique_size = function(units) {
+    ne_temp = matrix(ne_original[units,], ncol=ncol(ne_original))
+    sum(apply(ne_temp^2, 2, prod)) # the apply() gives 1 for assignments which all focals are connected to.
+  }
+  get_clique = function(units) {
+    ne_temp = matrix(ne_original[units,], ncol=ncol(ne_original))
+    a = apply(ne_temp^2, 2, prod) # =1 for assignments which all focals are connected to.
+    keep_ass = which(a==1)
+    return(ne_original[units, keep_ass])
+  }
+
+  while(CONT) {
+    # print(paste("Dim of NE=",dim(ne)))
+    units = as.numeric(rownames(ne))  # 2, 5, 6, ...
+    ass = as.numeric(colnames(ne))
+
+    # degree = as.numeric(rowSums(ne))
+    # rand_id = sample(1:length(units), 1, prob = degree) # random unit
+    rand_id = sample(1:length(units), 1) # select a random unit, only an index of that unit but not rowname.
+    focal_unit = units[rand_id]  # rowname of that randomly selected unit.
+
+    # print(paste("Select", focal_unit)) # unit name.
+    iass_remove = which(ne[rand_id,]==0)  # remove assignments that are not connected to this randomly selected unit.
+
+    ne = ne[-c(rand_id),-iass_remove]
+
+    focal_units = c(focal_units, focal_unit)  # add this unit to the set of focal units.
+
+    nsize = get_clique_size(focal_units)
+    if(nsize > num_ass) {
+      # print(paste("Added focal. Total", length(focal_units),"focal units. Clique size=", nsize,'\n'))
+    } else { # the algorithm ends if after adding this unit, num of assignments in the biclique is below num_ass
+      focal_units = setdiff(focal_units, focal_unit)
+      CONT = FALSE
+    }
+  }
+
+  focal_ass = as.numeric(colnames(ne))
+  clique = get_clique(focal_units)
+  return(list(clique=clique))
+}
+
+#' Decomposing Null Exposure Graph Using the Greedy Algorithm
+#'
+#' Outputs a biclique decomposition of the null-exposure graph using the greedy algorithm
+#' proposed above. The resulting decomposition paritions the assignment space,
+#' while the unit space may overlap.
+#'
+#' @param ne The null-exposure graph
+#' @param Zobs_id The index location of the observed assignment vector in \code{Z}, \code{Z_a}, and \code{Z_b}.
+#' @param num_ass Controls the minimum number of focal assignments in the biclique found
+#' by the algorithm (algorithm runtime is sensitive to this value).
+#' @param stop_at_Zobs A Boolean indicating whether the decomposition algorithm should stop when the clique containing the observed assignment is found.
+#' @return If \code{stop_at_Zobs} is \code{TRUE}, a matrix representing the clique to condition upon.
+#' If \code{stop_at_Zobs} is \code{FALSE}, a list containing the clique decomposition of the null-exposure graph.
+#' @export
+cliquedecom_greedy = function(ne, Zobs_id, num_ass, stop_at_Zobs){

```

Comment:
should we have "ne" be changed to "NEgraph" so that it matches the inputs to the other decom algo

Reply:
I've changed all "ne" to "NEgraph".

Reply:
Also unified the names of greedy algo and bimax algo, in the format of "out_clique_decomposition_bimax/greedy". Also changed the greedy decom to "out_greedy_decom"